HumbleKeysLibrary.cs

```diff
@@ -149,34 +342,78 @@ private Game ImportNewGame(Models.Order.TpkdDict.Tpk tpkd)
             {
                 gameInfo.Links.Add(MakeLink(tpkd?.gamekey));
             }
-
-            return PlayniteApi.Database.ImportGame(gameInfo, this);
+            // adds link to humble purchase
+            var links = new ObservableCollection<Link>();
+            if (UpdateStoreLinks(links, tpkd))
+            {
+                gameInfo.Links.AddRange(links.ToList());
+            }
+            PlayniteApi.Database.BeginBufferUpdate();
+            var game = PlayniteApi.Database.ImportGame(gameInfo, this);
+            if (groupTag != null)
+            {
+                game.TagIds.Add(groupTag.Id);
+                PlayniteApi.Database.Games.Update(game);
+            }
+            PlayniteApi.Database.EndBufferUpdate();
+            return game;
         }
 
-        private void UpdateExistingGame(Game existingGame, Order.TpkdDict.Tpk tpkd)
+        // If a game is expired, add tag 'Key: Unredeemable'
+        // If a game had been redeemed since last added to Playnite, remove the tag 'Key: Unredeemed' and add the tag 'Key: Redeemed'
+        // returns whether tags were updated or not 
+        bool UpdateRedemptionStatus(Game existingGame, Order.TpkdDict.Tpk tpkd, Tag groupTag = null)
         {
-            if (existingGame == null) { return; }
-            if (!Settings.keyTypeWhitelist.Contains(tpkd.key_type)) { return; }
+            var recordChanged = false;
+            if (existingGame == null) { return false; }
+            if (!Settings.keyTypeWhitelist.Contains(tpkd.key_type)) { return false; }
 
-            existingGame.Tags.RemoveAll(t => PAST_TAGS.Contains(t.Name));
+            // process tags on existingGame only if there was a change in tag status
 
-            existingGame.Tags.Add(new Tag( IsKeyPresent(tpkd) ? REDEEMED_STR : UNREDEEMED_STR ));
+            var existingRedemptionTagIds = existingGame.Tags?.Where(t => PAST_TAGS.Contains(t.Name)).ToList().Select(tag => tag.Id)??Enumerable.Empty<Guid>();
+            
+            // This creates a new Tag in the Tag Database if it doesn't already exist for 'Tag: Redeemed'
+            var tagIds = existingRedemptionTagIds.ToList();
+            var currentTagState = PlayniteApi.Database.Tags.Add(GetOrderRedemptionTagState(tpkd));
 
-            PlayniteApi.Database.Games.Update(existingGame);
+            if (groupTag != null)
+            {
+                if (existingGame.Tags != null && existingGame.Tags.All(tag => tag.Id != groupTag.Id))
+                {
+                    existingGame.TagIds.Add(groupTag.Id);
+                    recordChanged = true;
+                }
+            }
+            // existingGame already tagged with correct tag state
+            if (tagIds.Contains(currentTagState.Id)) return recordChanged;
+
+            // remove all tags related to key state
+            existingGame.TagIds.RemoveAll(tagId => tagIds.Contains(tagId));
+
+            existingGame.TagIds.Add(currentTagState.Id);
+
```

Comment:
Update() call is done outside of the function UpdateRedemptionStatus so that the BufferedUpdate triggers for the whole import process instead of an individual change.

Reply:
OK